ddl/multi_schema_change.go

```diff
@@ -56,18 +59,191 @@ func (d *ddl) MultiSchemaChange(ctx sessionctx.Context, ti ast.Ident) error {
+func isFinished(job *model.SubJob) bool {
+	return job.State == model.JobStateDone ||
+		job.State == model.JobStateRollbackDone ||
+		job.State == model.JobStateCancelled
+}
+
+func cloneFromSubJob(job *model.Job, sub *model.SubJob) *model.Job {
+	return &model.Job{
+		ID:              job.ID,
+		Type:            sub.Type,
+		SchemaID:        job.SchemaID,
+		TableID:         job.TableID,
+		SchemaName:      job.SchemaName,
+		State:           sub.State,
+		Warning:         sub.Warning,
+		Error:           nil,
+		ErrorCount:      0,
+		RowCount:        sub.RowCount,
+		Mu:              sync.Mutex{},
+		CtxVars:         sub.CtxVars,
+		Args:            sub.Args,
+		RawArgs:         sub.RawArgs,
+		SchemaState:     sub.SchemaState,
+		SnapshotVer:     sub.SnapshotVer,
+		RealStartTS:     job.RealStartTS,
+		StartTS:         job.StartTS,
+		DependencyID:    job.DependencyID,
+		Query:           job.Query,
+		BinlogInfo:      job.BinlogInfo,
+		Version:         job.Version,
+		ReorgMeta:       job.ReorgMeta,
+		MultiSchemaInfo: &model.MultiSchemaInfo{Revertible: sub.Revertible},
+		Priority:        job.Priority,
+		SeqNum:          job.SeqNum,
+	}
+}
+
+func mergeBackToSubJob(job *model.Job, sub *model.SubJob) {
+	sub.Revertible = job.MultiSchemaInfo.Revertible
+	sub.SchemaState = job.SchemaState
+	sub.SnapshotVer = job.SnapshotVer
+	sub.Args = job.Args
+	sub.State = job.State
+	sub.Warning = job.Warning
+	sub.RowCount = job.RowCount
+}
+
+func handleRevertibleException(job *model.Job, subJob *model.SubJob, err *terror.Error) {
+	if !isAbnormal(subJob) {
+		return
+	}
+	job.State = model.JobStateRollingback
+	job.Error = err
+	// Flush the cancelling state and cancelled state to sub-jobs.
+	for _, sub := range job.MultiSchemaInfo.SubJobs {
+		switch sub.State {
+		case model.JobStateCancelled:
+			if !sub.Revertible {
+				sub.State = model.JobStateCancelling
+			}
+		case model.JobStateRunning:
+			sub.State = model.JobStateCancelling
+		case model.JobStateNone, model.JobStateQueueing:
+			sub.State = model.JobStateCancelled
+		}
+	}
+}
+
+func isAbnormal(job *model.SubJob) bool {
+	return job.State == model.JobStateCancelling ||
+		job.State == model.JobStateCancelled ||
+		job.State == model.JobStateRollingback ||
+		job.State == model.JobStateRollbackDone
+}
+
+func appendToSubJobs(m *model.MultiSchemaInfo, job *model.Job) error {
+	err := fillMultiSchemaInfo(m, job)
 	if err != nil {
 		return err
 	}
+	m.SubJobs = append(m.SubJobs, &model.SubJob{
+		Type:        job.Type,
+		Args:        job.Args,
+		RawArgs:     job.RawArgs,
+		SchemaState: job.SchemaState,
+		SnapshotVer: job.SnapshotVer,
+		Revertible:  true,
+		CtxVars:     job.CtxVars,
+	})
+	return nil
+}
 
-	return checkAddColumnTooManyColumns(len(t.Cols()) + len(info.AddColumns) - len(info.DropColumns))
+func fillMultiSchemaInfo(info *model.MultiSchemaInfo, job *model.Job) (err error) {
+	switch job.Type {
+	case model.ActionAddColumn:
```

Comment:
only `ActionAddColumn` is support?

Reply:
For now, yes. To prevent this PR being too huge, only `ActionAddColumn` is supported.